src/interfaces/init.h

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERFACES_INIT_H
+#define BITCOIN_INTERFACES_INIT_H
+
+#include <fs.h>
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+struct NodeContext;
+
+namespace interfaces {
+class Base;
+class Echo;
+class IpcProcess;
+class IpcProtocol;
+
+//! Init interface providing access to other interfaces. The Init interface is
+//! the interface first exposed when a process spawns or connects to another
+//! process.
+//!
+//! The interfaces::Init interface is similar to other cross-process C++
+//! interfaces like interfaces::Node, interfaces::Wallet, interfaces::Chain and
+//! interfaces::ChainClient, providing virtual methods that can be called from
+//! other processes. What makes it special is that unlike other interfaces which
+//! are not implemented by every process—interfaces::Node is only implemented by
+//! the node process and interfaces::Wallet is only implemented by the wallet
+//! process—interfaces::Init is implemented by every process that supports being
+//! spawned, and it is the initial interface returned by the
+//! IpcProtocol::connect(fd) method, allowing the parent process to control the
+//! child process after the connection is established. The interfaces::Init
+//! interface has methods that allow the parent process to get access to every
+//! interface supported by the child process, and when the parent process frees
+//! the interfaces::Init object, the child process shuts down.
```

Comment:
I think this comment can be improved: 

* "Init interface providing access to other interfaces", this should underscores access is provided across process boundaries
* "providing virtual methods that can be called from other processes", AFAIU, this is confusing as interfaces::Init provides both process control method (e.g `makeEchoIpc()`, in the future `makeWallet`) and an interface `IpcProtocol` wrapping a libmultiprocess `ProxyClient/ProxyServer`. I would be better to document that both control and data exchange functions are supported by interfaces::Init

Overall, I still think `Init`/`LocalInit` are loosely-defining names. The names doesn't indicate what is initiated nor who is local and remote. `IPCSupervisor`/`IPCWrapper` as suggested would be better.

Reply:
I think my main concern is about lighting the interfaces _hierarchy_, where they're produced, and when they're consumed. See my graphical representation proposal as a way to improve it.

> Other reasonable names for interfaces::Init might be interfaces::Initial or interfaces::Factory or interfaces::Process or interfaces::Global.

+1 for `interfaces::Factory`.

> Maybe it could be called interfaces::InitImpl or interfaces::Startup or interfaces::LocalProcess

+1 for `interfaces::Startup`.

> Maybe it could be structured differently or not inherit from Init.

At least, I think `LocalInitImpl` which is different for `bitcoin-gui`/`bitcoin-node`/`bitcoin-wallet`,`bitcoind` should be prefixed like `GuiInit`, `NodeInit`, ...